### DISCUSSION TITLE
RuleObject里的attrs读取的是HTTP头里的字段吗？

### DISCUSSION THREAD

**Discussion by mlkt** [v2fly/v2ray-core#663] [Q&A] [answered]
[https://www.v2fly.org/config/routing.html#ruleobject](https://www.v2fly.org/config/routing.html#ruleobject) 文档里写的：
> attrs: string
> 
> （V2Ray 4.18+）一段脚本，用于检测流量的属性值。当此脚本返回真值时，此规则生效。
> 
> 脚本语言为 Starlark (opens new window)，它的语法是 Python 的子集。脚本接受一个全局变量 attrs，其中包含了流量相关的属性。
> 
> 目前只有 http 入站代理会设置这一属性。
> 
> 示例：
> 
> 检测 HTTP GET："attrs[':method'] == 'GET'"
> 检测 HTTP Path："attrs[':path'].startswith('/test')"
> 检测 Content Type："attrs['accept'].index('text/html') >= 0"

:method、:path、accept 这些值似乎指的是HTTP/2里的头信息，那如果传入请求用的是HTTP/1.1那是不是就不能判断url了？（因为HTTP/1.1的头里没有‘:path’），还是说attrs这个map的key是有一个预定义的列表？我看了下代码，没怎么看懂attrs这个变量的值来源，似乎是从pb直接反序列化得到的？

假如我的HTTP请求里有一个自定义头名叫'test'，
那我可以用 "attrs['test'] == 'yes'" 来判断这个头是否等于‘yes’吗？

**Comment by kslr** [comment, marked as answer]
是这里手动添加的 https://github.com/v2fly/v2ray-core/blob/master/proxy/http/server.go#L251

你也可以添加新 header

**Reply by mlkt** [reply]
噢噢，原来是这生成的，好的，多谢
